tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+use crate::loom::sync::atomic::AtomicU8;
+use crate::loom::sync::Mutex;
+use crate::util::linked_list::{self, LinkedList};
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::Ordering::SeqCst;
+use std::task::{Context, Poll, Waker};
+
+/// Notify a single task to wake up.
+///
+/// `Notify` provides a basic mechanism to notify a single task of an event.
+/// `Notify` itself does not carry any data. Instead, it is to be used to signal
+/// another task to perform an operation.
+///
+/// The synchronization details of `Notify` are similar to
+/// [`thread::park`][park] and [`Thread::unpark`][unpark] from std. A [`Notify`]
+/// value contains a single permit. [`Notify::recv`] waits for the permit to be
+/// made available, consumes the permit, and resumes. [`Notify::notify_one`]
+/// sets the permit, waking a pending task if there is one.
+///
+/// `Notify` can be thought of as a [`Semaphore`] starting with 0 permits.
+/// [`Notify::recv`] waits for a permit to become available and
+/// [`Notify::notify_one`] sets a permit **if there currently are no available
+/// permits**.
+///
+/// If `notify_one` is called **before** `recv()`, then the next call to
+/// `recv()` will complete immediately, consuming the permit. Any subsequent
+/// calls to `recv()` will wait for a new permit.
+///
+/// If `notify_one` is called **multiple** times before `recv()`, only a
+/// **single** permit is stored. The next call to `recv()` will complete
+/// immediately, but the one after will wait for a new permit.
+///
+/// # Examples
+///
+/// Basic usage.
+///
+/// ```
+/// use tokio::sync::Notify;
+/// use std::sync::Arc;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let notify = Arc::new(Notify::new());
+///     let notify2 = notify.clone();
+///
+///     tokio::spawn(async move {
+///         notify2.recv().await;
+///         println!("received a notification");
+///     });
+///
+///     notify.notify_one();
+/// }
+/// ```
+///
+/// Unbound mpsc channel.
+///
+/// ```
+/// use tokio::sync::Notify;
+/// use std::collections::VecDeque;
+/// use std::sync::Mutex;
+///
+/// struct Channel<T> {
```

Comment:
I love example! However it also raises a question: How can we use `Notify` to implement traits - e.g. a channel that is composed of `Stream` and `Sink`. It seems like those would need to require having a `poll` fn exposed instead of an `async fn`.

I actually like the `async fn` much more since we should somehow try to get away from using bare `Future`s  and poll functions. But the composability story with `async fn`s is unfortunately still not there.

Reply:
I'm... punting on that 😆 I'm placing the emphasis on `async fn` for this primitive. For `poll` implementations, there is `AtomicWaker`, which is more or less the same thing but with wakers...

Do you have a suggestion on how to move forward?

Reply:
I think one thing we need to find out for this is where it will be used 😀
Some APIs like the one in the example will definitely do well with waiting just bein an `async fn`. It's hard to say whether it will also be used for poll based stuff, or whether implementations of those traits will become more rare anyway.

I guess it might be fair to include it as proposed, and do API modifications on a future pre 1.0 release if deemed necessary.

Reply:
I like punting :)